OpenAPI v3

Jefftree opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

PR for OpenAPI v3. Based off the k/kube-openapi PR: https://github.com/kubernetes/kube-openapi/pull/247

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
/openapi/v3 endpoint will be populated with OpenAPI v3 if the feature flag is enabled
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/triage accepted

Jefftree committed:
Add OpenAPI v3 publishing under feature gate OpenAPIV3

Jefftree committed:
Generate openapi v3 static files

apelisse commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go:
```diff
@@ -218,6 +219,7 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
 		crdHandler,
 	)
 	openapiController := openapicontroller.NewController(s.Informers.Apiextensions().V1().CustomResourceDefinitions())
+	openapiv3Controller := openapiv3controller.NewController(s.Informers.Apiextensions().V1().CustomResourceDefinitions())
```
Are we doing this things unconditionally without regard for the feature gate value?

apelisse commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/builder.go:
```diff
@@ -74,6 +87,9 @@ type Options struct {
 	// Convert to OpenAPI v2.
 	V2 bool
 
+	// Used only by Server Side Apply. Create a V2 schema that includes fields from OpenAPI V3 for SSA to make better decisions
```
It's unclear what these better decisions are ;-)

sttts replied:
this needs much more explaination, especially because it leaks into other packages. What does this do *exactly*?

Jefftree replied:
The V2 flag is a parameter to feed the OpenAPI spec through the [ToStructuralOpenAPIV2](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/v2/conversion.go#L25) flag. 

It does three things: StripValueValidation, StripNullable, and unsets a field if XPreserveUnknownFields is true. I believe this last criteria is not strictly related to v2 but is [required by kubectl](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/v2/conversion.go#L76-L89)

Server side apply sets V2: false originally but that is a bit inaccurate because the output is still an OpenAPI v2 spec. Logically speaking V2: false should represent a v3 spec. SSA only set the flag to false because it doesn't want to go through the ToStructuralOpenAPIV2, specifically XPreserveUnknownFields:true fields to be nulled.

I think the ideal way to do this is to create a flag like StripTypesWithPreserveUnknownFields and completely delete the V2 flag (because it can be expressed as a combination of this flag, StripValueValidation, and StripNullable) but that creates a much bigger delta in the code and is more error prone.

I updated the flag to `OverrideToStructuralOpenAPIV2` and added a brief description of its purpose in the code. Let me know your thoughts, thanks!

apelisse commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapiv3/controller.go:
```diff
@@ -0,0 +1,295 @@
+	"k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder"
+)
+
+// Controller watches CustomResourceDefinitions and publishes OpenAPI v3
```
How much of this is different from the v2 controller?

Jefftree replied:
The same resources are watched, but the spec update process is completely different. The structs used to store the CRDs are also different (group/version mapping) and there is not much shared between v2 and v3.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -1361,7 +1361,7 @@ func buildOpenAPIModelsForApply(staticOpenAPISpec *spec.Swagger, crd *apiextensi
 	for _, v := range crd.Spec.Versions {
 		// Defaults are not pruned here, but before being served.
-		s, err := builder.BuildSwagger(crd, v.Name, builder.Options{V2: false, StripValueValidation: true, StripNullable: true, AllowNonStructural: false})
+		s, err := builder.BuildSwagger(crd, v.Name, builder.Options{V2: true, OverridePruneV2Fields: true, StripValueValidation: true, StripNullable: true, AllowNonStructural: false})
```
why those settings? This needs a go comment

Jefftree replied:
https://github.com/kubernetes/kubernetes/pull/105945#discussion_r746235701

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/builder.go:
```diff
@@ -65,6 +70,14 @@ var (
 )
 
+// The path for definitions in OpenAPI v2 and v3 are different. Translate the path if necessary
+func translateRefIfNeeded(schemaRef string, v2 bool) string {
```
refForOpenAPIVersion would be a better name.

sttts replied:
rule of thumb: use function names that don't force you to look into the function implementation to know what it does

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/builder.go:
```diff
@@ -172,8 +187,30 @@ func BuildSwagger(crd *apiextensionsv1.CustomResourceDefinition, version string,
+		return nil, err
+	}
+
+	openAPIv3Spec, err := builder3.BuildOpenAPISpec([]*restful.WebService{b.ws}, b.getOpenAPIConfig(false))
```
return builder3.BuildOpenAPISpec(...)

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/builder.go:
```diff
@@ -172,8 +187,30 @@ func BuildSwagger(crd *apiextensionsv1.CustomResourceDefinition, version string,
+}
+
+// BuildSwagger builds swagger for the given crd in the given version
+func BuildSwagger(crd *apiextensionsv1.CustomResourceDefinition, version string, opts Options) (*spec.Swagger, error) {
```
BuildOpenAPIV2

I think we have mostly banned the term Swagger from the code by now. If not, we should.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapiv3/controller.go:
```diff
@@ -0,0 +1,295 @@
+	c := &Controller{
+		crdLister:      crdInformer.Lister(),
+		crdsSynced:     crdInformer.Informer().HasSynced,
+		queue:          workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "crd_openapi_controller"),
```
use a new name, `crd_openapi_controller` is used already

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapiv3/controller.go:
```diff
@@ -0,0 +1,295 @@
+
+	// specs per version and per CRD name
+	lock           sync.Mutex
+	v3crdSpecsByGV map[string]map[string]*customResourceDefinitionData
```
specsByGV – everything here is v3 and crd.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapiv3/controller.go:
```diff
@@ -0,0 +1,295 @@
+	openAPIV3Service *handler3.OpenAPIService
+
+	// specs per version and per CRD name
+	lock           sync.Mutex
```
what is the lock for? I only see it being used in sync.

Jefftree replied:
Yeah, this is the same convention that is used in the openapi v2 crd controller. Prevent a race when modifying the CRD list.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapiv3/controller.go:
```diff
@@ -0,0 +1,295 @@
+// customResourceDefinitionData holds name and group/version information about a particular CRD spec
+type customResourceDefinitionData struct {
+	name     string
+	gvString string
```
any reason for this encoding of a GV? We have schema.GroupVersion. Can't we use that directly everywhere?

Jefftree replied:
This is a special encoding consisting of "apis/<group>/<version>" to make OpenAPI group updates easier for publishing.  I suppose we could change this to to use the schema.GroupVersionResource if that's cleaner

sttts replied:
use the GV directly. No need to encode as a string. https://goplay.space/#oPzh-rcpWFf

sttts commented on staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go:
```diff
@@ -51,3 +55,36 @@ func (oa OpenAPI) Install(c *restful.Container, mux *mux.PathRecorderMux) (*hand
+		klog.Fatalf("Failed to create OpenAPIService: %v", err)
+	}
+
+	err = openAPIVersionedService.RegisterOpenAPIV3VersionedService("/openapi/v3", mux)
```
Compare https://github.com/kubernetes/enhancements/pull/2898/files#r745380166. IMO we have the chance now to get rid of all the aggregator wiring by moving the spec into the group version handler.

sttts replied:
I think we could even get rid of the whole apiextensions spec generation controller, and just wire it cleanly into the customresourcedefinition handler, and generate on demand.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go:
```diff
@@ -17,6 +17,7 @@ limitations under the License.
 package aggregator
 
 import (
+	// "encoding/json"
```
remove

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/downloader.go:
```diff
@@ -95,12 +97,89 @@ func (s *Downloader) Download(handler http.Handler, etag string) (returnSpec *sp
+	Paths []string `json:"Paths"`
+}
+
+func (s *Downloader) DownloadV3(handler http.Handler, etag string) (returnSpec map[string]*spec3.OpenAPI, err error) {
```
I didn't expect a downloader at all. We moved to group-version instead of group alone in order to get rid of all this logic.

Jefftree replied:
https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/2896-openapi-v3/README.md#aggregator

This was the mechanism I described in the KEP. It's possible we can optimize it to avoid the download but I think we should resolve https://github.com/kubernetes/enhancements/pull/2898/files#r745380166 first.

Jefftree replied:
Per the discussion here and in https://github.com/kubernetes/enhancements/pull/2898/files#r745380166, I think proxying the requests instead of downloading is something that would be beneficial for performance. I think if we don't publish the `apis/g/v/openapi` route, we can still publish at `openapi/v3/apis/g/v` and bake a bit more proxy logic into the aggregator to avoid the download. 

If that's the path we'd like to take, would it be alright for this to be deferred to beta? This change should be 100% transparent to the user and is backwards compatible.

sttts commented:
/hold

I think we should reconsider our choice from the KEP and publish under `/apis/group/version/openapi/v3`. This will make us get rid of the aggregation layer completely, and
1. the aggregated APIServices will get that endpoint proxied automatically
2. apiextensions-apiserver can add a dynamic handler similar to discovery today and be totally distinct from kube-aggregator.

p0lyn0mial commented:
> I think we should reconsider our choice from the KEP and publish under /apis/group/version/openapi/v3. This will make us get rid of the aggregation layer completely, and ...

Given that the biggest users of OAS spec today are command-line clients and users create more and more CRDs in their clusters this would allow for scaling. It seems that we could create the spec dynamically without any caching layer.

Jefftree commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/builder.go:
```diff
@@ -486,13 +528,14 @@ func (b *builder) getOpenAPIConfig() *common.Config {
 			def[fmt.Sprintf("%s/%s.%s", b.group, b.version, b.kind)] = common.OpenAPIDefinition{
-				Schema: *b.schema,
+				Schema:       *b.schema,
+				Dependencies: []string{objectMetaType},
```
We used to serve an incomplete crd schema without the objectmeta reference, but it was never surfaced because the aggregator would include the objectmeta in during aggregation.

Jefftree commented:
> It seems that we could create the spec dynamically without any caching layer.

See https://github.com/kubernetes/kubernetes/pull/105945#discussion_r745886376, this should be independent of publishing under `/apis/group/version/openapi/v3`, as long as we are able to skip the download and aggregation performed by the aggregator, we can further optimize the openapi generation for crds.

I think we should have a more thorough discussion around this for beta.

Jefftree commented:
> /hold
> 
> I think we should reconsider our choice from the KEP and publish under `/apis/group/version/openapi/v3`. This will make us get rid of the aggregation layer completely, and
> 
> 1. the aggregated APIServices will get that endpoint proxied automatically
> 2. apiextensions-apiserver can add a dynamic handler similar to discovery today and be totally distinct from kube-aggregator.

Summarizing our discussion from here and slack:

1. For alpha, keep the downloader and aggregator 
2. For beta, revisit whether it's necessary to have an `openapi/v3` discovery endpoint at all. We can decouple the aggregator and apiextensions by proxying `openapi/v3/<group>/<version>` requests rather than downloading all the specs and serving them at the aggregator layer. This allows us to be more flexible and perhaps remove the need for the apiextensions spec generation controller and generate the spec on demand. We'll have better scaling for clusters with a large number of CRDs.

/unhold

DangerOnTheRanger committed:
Add integration tests for OpenAPI v3.

Jefftree commented:
/assign @apelisse @sttts @deads2k

Jefftree committed:
go mod vendor

apelisse commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapiv3/controller.go:
```diff
@@ -0,0 +1,295 @@
+	return nil
+}
+
+func (c *Controller) updateV3CRD(crd *apiextensionsv1.CustomResourceDefinition, name, versionName string, v3 *spec3.OpenAPI) error {
```
Can we rename that "v3" variable here? maybe `newSpec` or something?

sttts replied:
unresolved

apelisse commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go:
```diff
@@ -113,17 +119,16 @@ func BuildAndRegisterAggregator(downloader *Downloader, delegationTarget server.
 	if err != nil {
 		return nil, err
 	}
+	duration := time.Since(start)
```
Are we planning on having a similar counter for v3?

Jefftree replied:
This metric is for the duration for aggregation, we avoid aggregation in OpenAPI v3 so I'm not sure if it'll be useful at all?

apelisse commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go:
```diff
@@ -331,6 +409,16 @@ func (s *specAggregator) AddUpdateAPIService(handler http.Handler, apiService *v
 	s.rwMutex.Lock()
 	defer s.rwMutex.Unlock()
+	if utilfeature.DefaultFeatureGate.Enabled(features.OpenAPIV3) {
+		s.openAPIV3rwMutex.Lock()
```
Should we lock both mutex here? Any chance we could get into a deadlock situation?

Jefftree replied:
The v2 and v3 locks are for the v2 and v3 spec map respectively and are independent of each other. There isn't another place in the code where both locks are used so deadlock shouldn't be possible at the moment. We could do the writes one after the other to prevent both locks from being held at the same time but we won't be able to use the `defer` anymore.

apelisse replied:
The section at the beginning is only taking care of v3 and the section at the end is taking care of v2, can't you split in two? Also, I'd probably still use defer by splitting into two functions? I think this shows how the Aggregator might be better as two separate aggregator here.

Jefftree replied:
Completed refactored per sttts's suggestion.

apelisse commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/downloader.go:
```diff
@@ -101,6 +103,75 @@ func (s *Downloader) Download(handler http.Handler, etag string) (returnSpec *sp
+				if err != nil {
+					return nil, err
+				}
+				etag := writer.Header().Get("Etag")
```
What does this return if there's no etag? I can see above that you're comparing etags, but I suspect this returns an empty string if it doesn't exist, and empty strings will compare as equal.

apelisse replied:
On the other hand, since v3 will ALWAYS have etags enabled (compared to v2 where the feature was added later on), you can maybe assume that it's there?

Jefftree replied:
The etag should be guaranteed to exist if the APIService uses the OpenAPI v3 handlers that we provide. Is that a safe assumption to make? If the APIService is old and does not publish v3, then we don't care about the etag because `len(openAPIWriter.data)` will be 0.

apelisse commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/downloader.go:
```diff
@@ -101,6 +103,75 @@ func (s *Downloader) Download(handler http.Handler, etag string) (returnSpec *sp
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Accept", "application/json")
```
Should we use the `If-None-Match` that v2 uses? That would avoid the transfer altogether. Also possibly as an optimization later, we should use v3 since it's much much faster (assuming all of that doesn't go away when we proxy the request).

Jefftree replied:
Ah yes, that'll save quite a bit of transfers. Will add the header

apelisse commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/builder.go:
```diff
@@ -74,6 +87,12 @@ type Options struct {
 	// Convert to OpenAPI v2.
 	V2 bool
 
+	// Only takes effect if the flag and V2 and both set to true. If the condition is reached,
+	// publish OpenAPI V2 but skip running the spec through ToStructuralOpenAPIV2
+	// This prevents XPreserveUnknownFields:true fields from being cleared
+	// Used only by server side apply
+	OverrideToStructuralOpenAPIV2 bool
```
I'm not thrilled by a variable that strongly depends on another

apelisse replied:
it's unfortunate how a lot of these options only make sense for v2 and or to fit things into v2 and server-side apply. Hopefully we can get rid of this when we deprecate v2.

Jefftree commented:
/retest

apelisse commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -1361,7 +1361,8 @@ func buildOpenAPIModelsForApply(staticOpenAPISpec *spec.Swagger, crd *apiextensi
 		// Defaults are not pruned here, but before being served.
-		s, err := builder.BuildSwagger(crd, v.Name, builder.Options{V2: false, StripValueValidation: true, StripNullable: true, AllowNonStructural: false})
+		// See flag description in builder.go for flag usage
+		s, err := builder.BuildOpenAPIV2(crd, v.Name, builder.Options{V2: true, OverrideToStructuralOpenAPIV2: true, StripValueValidation: true, StripNullable: true, AllowNonStructural: false})
```
To be fair, I tried to look at it and I couldn't find a risky way to improve without refactoring a bunch. I'd hope we get rid of OpenAPI v2 rather than spend too much time refactoring v2/v3. WDYT?

Jefftree replied:
I'd like to completely get rid of the V2 field in the options in a follow up PR and refactor the code so the options make more sense. I'm not sure how well we can improve this section in this PR without going through the refactor.

apelisse replied:
Yeah, I'd love that, but I agree that it might be too ambitious for now.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/builder.go:
```diff
@@ -84,8 +103,7 @@ type Options struct {
 
-// BuildSwagger builds swagger for the given crd in the given version
-func BuildSwagger(crd *apiextensionsv1.CustomResourceDefinition, version string, opts Options) (*spec.Swagger, error) {
+func generateBuilder(crd *apiextensionsv1.CustomResourceDefinition, version string, opts Options) (*builder, error) {
```
a builder for a builder 🤷

apelisse replied:
Pray we don't alter it further with `GenericInstanceHandlerBuilderBuilder`

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -52,6 +52,7 @@ import (
 	openapibuilder "k8s.io/kube-openapi/pkg/builder"
```
rename to builder2

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/builder.go:
```diff
@@ -172,8 +190,26 @@ func BuildSwagger(crd *apiextensionsv1.CustomResourceDefinition, version string,
+		return nil, err
+	}
+
+	openAPISpec, err := openapibuilder.BuildOpenAPISpec([]*restful.WebService{b.ws}, b.getOpenAPIConfig(true))
```
return openapibuilder.BuildOpenAPISpec

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/builder.go:
```diff
@@ -348,26 +384,26 @@ func (b *builder) buildRoute(root, path, httpMethod, actionVerb, operationVerb s
 		// unknown fields for anything else.
 	} else {
-		if v2 {
+		if opts.V2 && !opts.OverrideToStructuralOpenAPIV2 {
```
OverrideToStructuralOpenAPIV2 is really `filterSchemaForKubectlOpenapiV2Validation`. Can we rename that?

Jefftree replied:
Renamed to `SkipFilterSchemaForKubectlOpenAPIV2Validation` because we want the inverse boolean value. Without inversing the boolean, the code would need to change in a lot more places (Changing SSA usage vs change all usage other than SSA)

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/builder.go:
```diff
@@ -348,26 +384,26 @@ func (b *builder) buildRoute(root, path, httpMethod, actionVerb, operationVerb s
 	// adding additionalProperties=true support to explicitly allow additional fields.
 	// TODO: fix kubectl to understand additionalProperties=true
-	if schema == nil || (v2 && (schema.XPreserveUnknownFields || crdPreserveUnknownFields)) {
+	if schema == nil || ((opts.V2 && !opts.OverrideToStructuralOpenAPIV2) && (schema.XPreserveUnknownFields || crdPreserveUnknownFields)) {
```
this change I don't understand. `OverrideToStructuralOpenAPIV2` is supposed to only have an effect if `V2` is true. So why do we care here? Can we remove this change?

Jefftree replied:
v2 is true here. The code path here clears the schema if XPreservedUnknownFields is set which is not what SSA wants.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapiv3/controller.go:
```diff
@@ -0,0 +1,295 @@
+	"k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder"
+)
+
+// Controller watches CustomResourceDefinitions and publishes OpenAPI v3
```
publish where?

Jefftree replied:
to the `/openapi/v3` and `/openapi/v3/<group>/<version>` handler of the apiextensions apiserver. It's the same comment as the openapi v2 controller, is an explicit path necessary?

sttts replied:
I think I missed that it publishes to the `OpenAPIService` that is passed in the `Run` func

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapiv3/controller.go:
```diff
@@ -0,0 +1,295 @@
+
+	if !cache.WaitForCacheSync(stopCh, c.crdsSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
```
does this mean we hide an error?

Jefftree replied:
I think this is the best we can do since the controller runs in its own goroutine? The V2 controller also does this: 
 https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go#L97-L100

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapiv3/controller.go:
```diff
@@ -0,0 +1,295 @@
+		c.crdSpecsByGV[gvString] = map[string]*customResourceDefinitionData{}
+	}
+
+	prevCRD, ok := c.crdSpecsByGV[gvString][name]
```
oldSpec. This is not a CRD.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapiv3/controller.go:
```diff
@@ -0,0 +1,295 @@
+				needOpenAPIUpdate = true
+				delete(crdList, crdName)
+			}
+		}
```
```
_, needOpenAPIUpdate  := crdList[name]
if needOpenAPIUpdate {
  delete(crdList, name)
}
```

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapiv3/controller.go:
```diff
@@ -0,0 +1,295 @@
+			}
+		}
+		if needOpenAPIUpdate {
+			if len(crdList) == 0 {
```
this can never happen

Jefftree replied:
If there are no crds left in a group/version, the group/version should be deleted.

Jefftree replied:
The variable naming wasn't very clear. I've updated the name of this to crdListByGV so it's clear that this is only checking for the scenario of deleting the last CRD in a group rather than deleting all CRDs.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapiv3/controller.go:
```diff
@@ -0,0 +1,295 @@
+
+	if err != nil {
+		return err
+	}
```
remove

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapiv3/controller.go:
```diff
@@ -0,0 +1,295 @@
+		return err
+	}
+
+	if c.openAPIV3Service != nil {
```
in line 203 you would crash with openAPIV3Service==nil

Jefftree replied:
Ah this is slightly outdated code when the v3 code was not in its own controller. 

With the openapi v3 code in its own controller, openAPIV3Service should be guaranteed to be non nil and I've removed this check.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapiv3/controller.go:
```diff
@@ -0,0 +1,295 @@
+
+	// specs per version and per CRD name
+	lock         sync.Mutex
+	crdSpecsByGV map[string]map[string]*customResourceDefinitionData
```
drop "crd". Everything is about CRDs here.

specsByGVandName

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapiv3/controller.go:
```diff
@@ -0,0 +1,295 @@
+
+// customResourceDefinitionData holds name and group/version information about a particular CRD spec
+type customResourceDefinitionData struct {
+	name     string
```
drop the name here. It is the key already of the map where this is used.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapiv3/controller.go:
```diff
@@ -0,0 +1,295 @@
+	gvString string
+	version  string
+	spec     *spec3.OpenAPI
+}
```
- gvString here is never used.
- version is never used.
- 
Only `*spec3.OpenAPI` remains. Get rid of the whole struct.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapiv3/controller.go:
```diff
@@ -0,0 +1,295 @@
+/*
+Copyright 2019 The Kubernetes Authors.
```
2021

sttts commented on staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go:
```diff
@@ -34,7 +36,9 @@ type OpenAPI struct {
-	spec, err := builder.BuildOpenAPISpec(c.RegisteredWebServices(), oa.Config)
+	w := c.RegisteredWebServices()
+
+	spec, err := builder.BuildOpenAPISpec(w, oa.Config)
```
unneccessary change

sttts replied:
rename import to builder2

sttts commented on staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go:
```diff
@@ -34,7 +36,9 @@ type OpenAPI struct {
 
 // Install adds the SwaggerUI webservice to the given mux.
 func (oa OpenAPI) Install(c *restful.Container, mux *mux.PathRecorderMux) (*handler.OpenAPIService, *spec.Swagger) {
```
rename to Install2

sttts commented on staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go:
```diff
@@ -51,3 +55,36 @@ func (oa OpenAPI) Install(c *restful.Container, mux *mux.PathRecorderMux) (*hand
 }
+
+func (oa OpenAPI) InstallV3(c *restful.Container, mux *mux.PathRecorderMux) *handler3.OpenAPIService {
+	w := c.RegisteredWebServices()
```
any need to have that that far up? Just use `range c.RegisteredWebServices()` below.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go:
```diff
@@ -51,3 +55,36 @@ func (oa OpenAPI) Install(c *restful.Container, mux *mux.PathRecorderMux) (*hand
+		spec, err := builder3.BuildOpenAPISpec(ws, oa.Config)
+		if err != nil {
+			klog.Errorf("Failed to build OpenAPI v3 for group %s, %q", gv, err)
+
```
remove empty line

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go:
```diff
@@ -74,9 +79,10 @@ func (s *specAggregator) GetAPIServiceNames() []string {
+	config *common.Config, pathHandler common.PathHandlerByGroupVersion) (SpecAggregator, error) {
 	s := &specAggregator{
-		openAPISpecs: map[string]*openAPISpecInfo{},
+		openAPISpecs:   map[string]*openAPISpecInfo{},
```
openAPIV2Specs

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go:
```diff
@@ -132,21 +137,89 @@ func BuildAndRegisterAggregator(downloader *Downloader, delegationTarget server.
 	return s, nil
 }
 
+func (s *specAggregator) UpdateOpenAPIV3(downloader Downloader, apiServiceName string, handler http.Handler) error {
```
iit's not apparent what an `Update*` func does. Is it adding another handler? Name it properly.

Jefftree replied:
Renamed to AddUpdateAPIServiceForOpenAPIV3 and added description

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go:
```diff
@@ -42,6 +46,7 @@ type SpecAggregator interface {
 	AddUpdateAPIService(handler http.Handler, apiService *v1.APIService) error
 	UpdateAPIServiceSpec(apiServiceName string, spec *spec.Swagger, etag string) error
 	RemoveAPIServiceSpec(apiServiceName string) error
+	UpdateOpenAPIV3(downloader Downloader, apiService string, handler http.Handler) error
```
This looks very odd. All other funcs are SomethingAPIServiceSpec. Why is this different? document what the func does.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go:
```diff
@@ -132,21 +137,89 @@ func BuildAndRegisterAggregator(downloader *Downloader, delegationTarget server.
+	// mutex protects all openAPI v3 members of this struct.
+	openAPIV3rwMutex sync.RWMutex
+
+	openAPIV3Specs map[string]map[string]*openAPIV3SpecInfo
```
spec by what?

sttts replied:
By usage I guess `ByGroupVersionAndGroup`, very odd. Why twice by group?

Jefftree replied:
APIService and then by GroupVersion. Updated the documentation

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go:
```diff
@@ -331,6 +409,16 @@ func (s *specAggregator) AddUpdateAPIService(handler http.Handler, apiService *v
+			}
+			delete(s.openAPIV3Specs, apiServiceName)
+		}
+		s.openAPIV3rwMutex.Unlock()
```
this is dangerous. Put the unlock into a defer. Otherwise, we risk a deadlock when `DeleteGroupVersion` panics.

Jefftree commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/builder.go:
```diff
@@ -65,6 +70,14 @@ var (
+		return schemaRef
+	}
+	return strings.Replace(schemaRef, definitionPrefix, v3DefinitionPrefix, 1)
+}
```
Added documentation for this

apelisse commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/controllerv3.go:
```diff
@@ -0,0 +1,183 @@
+/*
```
Did you need to call this file `controllerv3.go` since it's in the `openapiv3` package?

Jefftree replied:
ok, keeping as `controller.go`

apelisse commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/aggregator.go:
```diff
@@ -0,0 +1,200 @@
+		handler:    handler,
+		specs:      make(map[string]*openAPIV3SpecInfo),
+	}
+	return nil
```
does this need to return an error?

Jefftree replied:
hmm looks like it doesn't. In removing an apiservice could have failed if the newly aggregated spec doesn't build properly, but with v3 we simply delete an entry in a map and it's impossible to have an error.

apelisse replied:
We can change the function signature then

apelisse commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/controllerv3.go:
```diff
@@ -0,0 +1,183 @@
+
+// AddAPIService adds a new API Service to OpenAPI Aggregation.
+func (c *AggregationController) AddAPIService(handler http.Handler, apiService *v1.APIService) {
+	klog.Infof("ARST: %s", apiService.Name)
```
Are these (this one and the one below) for debugging purposes?

apelisse reviewed (commented):
Thanks, looks good to me.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -1361,7 +1361,8 @@ func buildOpenAPIModelsForApply(staticOpenAPISpec *spec.Swagger, crd *apiextensi
 		// Defaults are not pruned here, but before being served.
-		s, err := builder.BuildSwagger(crd, v.Name, builder.Options{V2: false, StripValueValidation: true, StripNullable: true, AllowNonStructural: false})
+		// See flag description in builder.go for flag usage
+		s, err := builder.BuildOpenAPIV2(crd, v.Name, builder.Options{V2: true, SkipFilterSchemaForKubectlOpenAPIV2Validation: true, StripValueValidation: true, StripNullable: true, AllowNonStructural: false})
```
SkipFilterSchemaForKubectlOpenAPIV2Validation is much better :+1:

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapiv3/controller.go:
```diff
@@ -0,0 +1,272 @@
+	}
+
+	var specs []*spec3.OpenAPI
+	for _, crd := range c.specsByGVandName[gv] {
```
nit: s/crd/spec/

sttts commented on staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go:
```diff
@@ -51,3 +53,33 @@ func (oa OpenAPI) Install(c *restful.Container, mux *mux.PathRecorderMux) (*hand
+			klog.Errorf("Failed to build OpenAPI v3 for group %s, %q", gv, err)
+
+		}
+		openAPIVersionedService.UpdateGroupVersion(gv, spec)
```
which GVs does grouped include here? Are these supposed to be the static groups in kube-apiserver? Does this mean we have to call this func late in the startup? What about dynamic APIServices?

Jefftree replied:
Yes, this is the list of all static GVs in kube-apiserver. Both this function and InstallV2 are called in `PrepareRun` of the apiserver.

Dynamic APIServices are added in the aggregator (same place as for openapi v2)

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/aggregator.go:
```diff
@@ -0,0 +1,204 @@
+	"k8s.io/kube-openapi/pkg/spec3"
+)
+
+// SpecAggregator calls out to http handlers of APIServices and merges specs. It keeps state of the last
```
s/merges/caches/

sttts replied:
`// TODO: remove the downloading and caching and proxy directly to the APIServices. This is possible because we don't have to merge here, which is cpu intensive in v2`

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/aggregator.go:
```diff
@@ -0,0 +1,204 @@
+	// mutex protects all members of this struct.
+	rwMutex sync.RWMutex
+
+	// OpenAPI V3 specs by APIService and then by group/version
```
typo

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/aggregator.go:
```diff
@@ -0,0 +1,204 @@
+type openAPIV3APIServiceInfo struct {
+	apiService v1.APIService
+	handler    http.Handler
+	specs      map[string]*openAPIV3SpecInfo
```
why a map? Isn't there just one per APIService?

Jefftree replied:
The apiextensions apiserver can publish multiple groups so I just adopted that kind of format here. For other apiservices, yes it's just one per.

sttts commented:
The factoring is so much better than a day or two ago. Very well done.

Without the feature gate being on, the new controllers and spec aggregator are not instantiated. So risk is minimal, as it should be with a feature-gates alpha implementation.

Jefftree commented:
/test pull-kubernetes-unit
(unrelated failure)

sttts commented:
/lgtm
/approve

Jefftree commented:
Rebased and regenerated the OpenAPI files to resolve a conflict. Code is unchanged

deads2k commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapiv3/controller.go:
```diff
@@ -0,0 +1,272 @@
+		}
+	}
+
+	// only start one worker thread since its a slow moving API
```
based on my read, I think we also rely on this to avoid some need for locks.  I think it's reasonable for this controller.

deads2k commented:
> Without the feature gate being on, the new controllers and spec aggregator are not instantiated. So risk is minimal, as it should be with a feature-gates alpha implementation.

Agree

/approve

deads2k commented on hack/update-openapi-spec.sh:
```diff
@@ -96,10 +96,21 @@ if ! kube::util::wait_for_url "https://${API_HOST}:${API_PORT}/healthz" "apiserv
 
 curl -w "\n" -kfsS -H 'Authorization: Bearer dummy_token' "https://${API_HOST}:${API_PORT}/openapi/v2" | jq -S '.info.version="unversioned"' > "${OPENAPI_ROOT_DIR}/swagger.json"
 
+kube::log::status "Updating " "${OPENAPI_ROOT_DIR}/v3 for OpenAPI v3"
```
I've often wondered why we produce these.  Do we really want to generate these?  I'd be happy to see them removed and/or generated from the source instead of the server.

apelisse replied:
I think it'd be great if we could generate from the source rather than the server, on the other hand, it's super useful to be able to see what changed or that nothing changed.

Jefftree replied:
I don't have a strong opinion about this and am happy to remove if needed. (It'll reduce the likelihood of merge conflicts as well). 

OTOH the `swagger.json` generated for v2 is used in some tests, but v3 has no such dependency.

thockin replied:
+1 for direction

roycaihw replied:
v2 spec is used for [generating clients](https://github.com/kubernetes-client/gen/blob/291f9a6ca5004cc91cfc4a744abf129c639d5682/openapi/preprocess_spec.py#L517-L521). Potentially we want to use a v3 equivalent in the future

apelisse commented:
/lgtm

thockin commented:
Mechanically LGTM.  I'll approve when lgtm by code-level owners

thockin commented:
@liggitt @dims @cheftako 2nd such tag this morning.  Sorry.

dims commented:
/approve

sttts commented:
/approve

to retrigger bot. @dims is supposed to tag legacy cloud providers. Odd.

Jefftree commented:
/priority important-soon

retrigger prow

Jefftree commented:
/approve

retrigger the bot

Jefftree commented:
There are a couple other OpenAPI changes that will cause merge conflicts:
https://github.com/kubernetes/kubernetes/pull/105057
https://github.com/kubernetes/kubernetes/pull/106154

Hoping these are merged sequentially and rebased

deads2k commented:
/milestone 1.23

deads2k commented:
/milestone v1.23

Jefftree commented:
> There are a couple other OpenAPI changes that will cause merge conflicts: #105057 #106154
> 
> Hoping these are merged sequentially and rebased

Both these PRs merged. Rebased off the latest master and regenerated the OpenAPI spec.

fedebongio commented:
/lgtm

apelisse commented:
/lgtm

Jefftree commented:
/retest

pacoxu commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/builder.go:
```diff
@@ -424,40 +456,46 @@ func addEmbeddedProperties(s *spec.Schema, v2 bool) {
+func generateBuildDefinitionsFunc(v2 bool) func() {
+	return func() {
+		namer = openapi.NewDefinitionNamer(runtime.NewScheme())
+		definitions = utilopenapi.GetOpenAPIDefinitionsWithoutDisabledFeatures(generatedopenapi.GetOpenAPIDefinitions)(func(name string) spec.Ref {
```
```suggestion
		definitions = utilopenapi.GetOpenAPIDefinitionsWithoutDisabledFeatures(generatedopenapi.GetOpenAPIDefinitions)(func(name string) spec.Ref {
```

pacoxu replied:
To fix the gofmt 😄

Jefftree replied:
This is an outdated comment right? The verify failure was before the last force push

jiahuif replied:
Could you point out what got changed?

jiahuif replied:
Oh I see. already outdated.

pacoxu replied:
Oh, outdated. Sorry.

Jefftree commented:
/retest
(flake)
